src/internal/connector/exchange/container_resolver.go

```diff
@@ -40,50 +48,147 @@ type containersEnumerator interface {
+
+	c, ok := cr.cache[folderID]
+	if !ok {
+		return nil, nil, clues.New("folder not cached").WithClues(ctx)
+	}
+
+	p := c.Path()
+	if p == nil {
+		return nil, nil, clues.New("folder has no path").WithClues(ctx)
+	}
+
+	return p, c.Location(), nil
+}
+
+func (cr *containerResolver) refreshContainer(
+	ctx context.Context,
+	id string,
+) (graph.CachedContainer, bool, error) {
+	ctx = clues.Add(ctx, "refresh_container_id", id)
+	logger.Ctx(ctx).Debug("refreshing container")
+
+	if cr.refresher == nil {
+		return nil, false, clues.New("nil refresher")
+	}
+
+	c, err := cr.refresher.refreshContainer(ctx, id)
+	if err != nil && graph.IsErrDeletedInFlight(err) {
+		logger.Ctx(ctx).Debug("container deleted")
+		return nil, true, nil
+	} else if err != nil {
+		// This is some other error, just return it.
+		return nil, false, clues.Wrap(err, "refreshing container").WithClues(ctx)
+	}
+
+	return c, false, nil
 }
 
 func (cr *containerResolver) idToPath(
 	ctx context.Context,
 	folderID string,
 	depth int,
-) (*path.Builder, *path.Builder, error) {
+) (*path.Builder, *path.Builder, bool, bool, error) {
 	ctx = clues.Add(ctx, "container_id", folderID)
 
 	if depth >= maxIterations {
-		return nil, nil, clues.New("path contains cycle or is too tall").WithClues(ctx)
+		return nil, nil, false, false, clues.New("path contains cycle or is too tall").WithClues(ctx)
 	}
 
 	c, ok := cr.cache[folderID]
 	if !ok {
-		return nil, nil, clues.New("folder not cached").WithClues(ctx)
+		c, shouldDelete, err := cr.refreshContainer(ctx, folderID)
+		if err != nil {
+			return nil, nil, false, false, clues.Wrap(err, "fetching uncached container")
+		}
+
+		if shouldDelete {
+			logger.Ctx(ctx).Debug("fetching uncached folder showed it was deleted")
+			return nil, nil, false, shouldDelete, err
+		}
+
+		if err := cr.addFolder(c); err != nil {
+			return nil, nil, false, false, clues.Wrap(err, "adding new folder").WithClues(ctx)
+		}
+
+		// Retry populating this container's paths.
+		// TODO(ashmrtn): May want to bump the depth here just so we don't get stuck
+		// retrying too much if for some reason things keep moving around?
+		pth, loc, _, shouldDelete, err := cr.idToPath(ctx, folderID, depth)
+		if err != nil {
+			err = clues.Wrap(err, "retry populating uncached folder")
+		}
+
+		return pth, loc, false, shouldDelete, err
 	}
 
 	p := c.Path()
 	if p != nil {
-		return p, c.Location(), nil
+		return p, c.Location(), true, false, nil
 	}
 
-	parentPath, parentLoc, err := cr.idToPath(
+	parentPath, parentLoc, parentCached, shouldDelete, err := cr.idToPath(
 		ctx,
 		ptr.Val(c.GetParentFolderId()),
 		depth+1)
 	if err != nil {
-		return nil, nil, clues.Wrap(err, "retrieving parent folder")
+		return nil, nil, true, false, clues.Wrap(err, "retrieving parent folder")
+	}
+
+	if !parentCached {
+		logger.Ctx(ctx).Debug("parent folder was refreshed")
+
+		newContainer, currentShouldDelete, err := cr.refreshContainer(ctx, folderID)
```

Comment:
not entirely happy with this suggestion. But "currentShouldDelete" is a mouthful, and makes it seem like we're inside a loop here.
```suggestion
		pc, pShouldDelete, err := cr.refreshContainer(ctx, folderID)
```